pims/base_frames.py

```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+        The ndarray that is returned by get_frame has the same dimension order
+        as the order of aggregate. """
+        return [d.name for d in self._dims if d.aggregate]
+
+    @aggregate.setter
+    def aggregate(self, value):
+        for dim_aggregate in value:
+            if dim_aggregate.lower() in ['x', 'y']:
+                raise ValueError('Aggregate does not take dimensions x or y.')
+            if dim_aggregate not in self.dims:
+                raise ValueError(('Dimension named ''{}'' does not exist ' +
+                                  'in this image.').format(dim_aggregate))
+
+        new_dims_aggr = [0] * len(value)
+        new_dims_rest = []
+        for dim in self._dims:
+            for n_aggr, dim_aggregate in enumerate(value):
+                if dim.name == dim_aggregate:
+                    new_dims_aggr[n_aggr] = dim
+                    dim.aggregate = True
+                    break
+            else:
+                new_dims_rest.append(dim)
+                dim.aggregate = False
+        self._dims = new_dims_rest + new_dims_aggr
+
+    @property
+    def iterate(self):
+        """ This determines which dimensions will be iterated over by the
+        FramesSequence. The last element will iterate fastest. """
+        return [d.name for d in self._dims if d.iterate]
+
+    @iterate.setter
+    def iterate(self, value):
+        for dim_iterate in value:
+            if dim_iterate.lower() in ['x', 'y']:
+                raise ValueError('Iterate does not take dimensions x or y.')
+            if dim_iterate not in self.dims:
+                raise ValueError(("Dimension named '{}' does not exist " +
+                                  "in this image.").format(dim_iterate))
+
+        new_dims_iter = [0] * len(value)
+        new_dims_rest = []
+        for dim in self._dims:
+            for n_iter, dim_iterate in enumerate(value):
+                if dim.name == dim_iterate:
+                    new_dims_iter[n_iter] = dim
+                    dim.iterate = True
+                    break
+            else:
+                new_dims_rest.append(dim)
+                dim.iterate = False
+        self._dims = new_dims_iter + new_dims_rest
+
+    @abstractmethod
+    def get_frame_2D(self, **ind):
+        """ The actual frame reader, defined by the subclassed reader.
+
+        This method should take exactly one keyword argument per dimension,
+        reflecting the index along each dimension. It returns a two dimensional
+        ndarray with shape `frame_shape_2D` and dtype `pixel_type`.
+        """
+        pass
+
+    def get_frame(self, i):
+        """ Returns a Frame of shape deterimend by aggregate. The property
+        iterate (together with the dimension default index) determines which.
+        multidimensional index is returned. """
+        # identify the indices to take along each dimension
+        i_prev = 0
```

Comment:
Agreed, but this is a quite tricky part of the code. I now have this:

        # list sizes of iterate dimensions
        iter_sizes = [self._sizes[k] for k in self._iterate]
        # list how much i has to increase to get an increase of coordinate n
        iter_cumsizes = np.append(np.cumprod(iter_sizes[::-1])[-2::-1], 1)
        # calculate the coordinates and update the coords dictionary
        iter_coords = (i // iter_cumsizes) % iter_sizes